tests/test-dirs/stack-or-heap.t/run.t

```diff
@@ -591,3 +599,35 @@ how to produce valid json.
   
   "could be stack or heap"
   
+  |  let z = Some x in
+  |               ^
+  
+  |  let z = Some x in
+  |          ^^^^^^
+  
+  "could be stack or heap"
+  
+  |  let z = Some x in
+  |               ^
+  
+  |  let z = Some x in
+  |          ^^^^^^
+  
+  "stack"
+  
+  |  let z = Some x in
+  |               ^
+  
+  |  let z = Some x in
+  |          ^^^^^^
+  
+  "could be stack or heap"
```

Comment:
This is unfortunate -- this really should be `stack`. The source structure-item is complete. But it's in a module that's incomplete. So `optimise_allocations` isn't run. If maybe we iterated through each structure-item one at a time, instead of calling `type_implementation` directly, and then run `optimise_allocations` each time, we'd get better output here. But that approach would have its own problems, because it means that weak polymorphism wouldn't work.

I don't propose making that change in this PR (or anytime soon, really). But I do think it's worth making both adding a comment to the test saying this is not as good as it could be (and a brief explanation of why) an internal ticket to note that we can do better here.